### PR TITLE
Allow for an optional client with react-query 

### PR DESCRIPTION
**Pull request by adairrr**
This change makes it so the generated react-query files can still compile even if the contract client is undefined as the component renders. It also maintains the developer use of the `enabled` flag ([docs](https://react-query-v3.tanstack.com/guides/dependent-queries)) for `useQuery` so that it may be enabled based on any custom boolean.

This change *still* mandates the presence of the client because the query is disabled as long as the client does not exist. 

The `XXXResponse | undefined` was added to make the typechecker happy but *alternatively*, we can avoid this by inserting a non-null bang after `client` in the function call (instead of the ternary) because the query is not enabled unless the client is present.

**Before**:
```ts
export interface Sg721CollectionInfoQuery {
  client: Sg721QueryClient;
  options?: UseQueryOptions<CollectionInfoResponse, Error, CollectionInfoResponse, (string | undefined)[]>;
}
export function useSg721CollectionInfoQuery({
  client,
  options
}: Sg721CollectionInfoQuery) {
  return useQuery<CollectionInfoResponse, Error, CollectionInfoResponse, (string | undefined)[]>(["sg721CollectionInfo", client.contractAddress], () => client.collectionInfo(), options);
}
```

**After**:
```ts
export interface Sg721CollectionInfoQuery {
  client?: Sg721QueryClient; // optional
  options?: UseQueryOptions<CollectionInfoResponse | undefined, Error, CollectionInfoResponse, (string | undefined)[]>;
}
export function useSg721CollectionInfoQuery({
  client,
  options
}: Sg721CollectionInfoQuery) {
  return useQuery<CollectionInfoResponse | undefined, Error, CollectionInfoResponse, (string | undefined)[]>(["sg721CollectionInfo", client?.contractAddress], () => client ? client.collectionInfo() : undefined, {
    ...options,
    enabled: !!client && options?.enabled, // ensure the client is present
  });
}
```

### CODE REVIEW

**Comment by pyramation** [comment]
awesome work @adairrr!

Can we make this an optional so we can merge?

- [ ] add `optionalClient`  after the contract name option

```js
{
            type: 'confirm',
            name: 'optionalClient',
            message: 'optionalClient?',
            default: false
 }
```

https://github.com/CosmWasm/cosmwasm-typescript-gen/blob/master/packages/cosmwasm-typescript-gen/src/commands/react-query.ts#L22


- [ ] add option to `react-query` method https://github.com/CosmWasm/cosmwasm-typescript-gen/blob/master/packages/cosmwasm-typescript-gen/src/react-query.ts#L11

- [ ] update the `wasm-ast-types` to accept the `optionalClient` flag

**Comment by pyramation** [comment]
also, can you remove the updates to the package versions? This will end up causing conflicts, and `lerna` will update those when we publish.

File: packages/wasm-ast-types/src/react-query.ts [hunk 1]
```diff
@@ -115,28 +130,53 @@ export const createReactQueryHook = ({
...
+                                            ),
+                                            t.optionalMemberExpression(
+                                                t.identifier('options'),
+                                                t.identifier('enabled'),
+                                                false,
+                                                true
+                                            )
+                                        )),
+                                    t.spreadElement(t.identifier('options'))
+                                ]),
```
Comment (adairrr) [inline packages/wasm-ast-types/src/react-query.ts:178]: Can clean this up if desired!

**Comment by adairrr** [comment]
> awesome work @adairrr!
> 
> Can we make this an optional so we can merge?
> 
> * [x]  add `optionalClient`  after the contract name option
> 
> ```js
> {
>             type: 'confirm',
>             name: 'optionalClient',
>             message: 'optionalClient?',
>             default: false
>  }
> ```
> 
> https://github.com/CosmWasm/cosmwasm-typescript-gen/blob/master/packages/cosmwasm-typescript-gen/src/commands/react-query.ts#L22
> 
> * [x]  add option to `react-query` method https://github.com/CosmWasm/cosmwasm-typescript-gen/blob/master/packages/cosmwasm-typescript-gen/src/react-query.ts#L11

> * [x]  update the `wasm-ast-types` to accept the `optionalClient` flag

Great suggestions, thank you! I also made the options an object so that extending them in the future won't be too much of a hassle.

File: packages/wasm-ast-types/src/react-query.ts (hunk 1 above)
Comment (adairrr) [reply packages/wasm-ast-types/src/react-query.ts:178]: Right now, it generates: `enabled: !!client && options?.enabled,` even if the client is definitely there.

Comment (sascha1337) [reply packages/wasm-ast-types/src/react-query.ts:178]: AST GANG <3

File: packages/wasm-ast-types/src/__snapshots__/react-query.spec.ts.snap [hunk 2]
```diff
@@ -9,7 +9,10 @@ export function useSg721CollectionInfoQuery({
   client,
   options
 }: Sg721CollectionInfoQuery) {
-  return useQuery<CollectionInfoResponse, Error, CollectionInfoResponse, (string | undefined)[]>([\\"sg721CollectionInfo\\", client.contractAddress], () => client.collectionInfo(), options);
+  return useQuery<CollectionInfoResponse, Error, CollectionInfoResponse, (string | undefined)[]>([\\"sg721CollectionInfo\\", client.contractAddress], () => client.collectionInfo(), {
+    enabled: !!client && options?.enabled,
+    ...options
```
Comment (adairrr) [inline packages/wasm-ast-types/src/__snapshots__/react-query.spec.ts.snap:14]: This is backwards; should read:
```
...options
enabled: !!client && options?.enabled,
```

**Comment by adairrr** [comment]
Ah, and would love these commits to be squashed rather than merged! No sense dirtying the repo.

File: packages/wasm-ast-types/src/react-query.ts [hunk 3]
```diff
@@ -115,28 +127,53 @@ export const createReactQueryHook = ({
...
+                                                t.unaryExpression('!', t.identifier('client'))
+                                            ),
+                                            t.optionalMemberExpression(
+                                                t.identifier('options'),
+                                                t.identifier('enabled'),
+                                                false,
+                                                true
+                                            )
+                                        )),
+                                ]),
```
Comment (pyramation) [inline packages/wasm-ast-types/src/react-query.ts:170]: @adairrr I'm guessing this should also be something like 

```js
!!options?.optionalClient ? /* new code you wrote */ : t.identifier('options')
```

File: packages/wasm-ast-types/src/react-query.ts [hunk 4]
```diff
@@ -172,22 +209,25 @@ interface ReactQueryHookQueryInterface {
     hookParamsTypeName: string;
     responseType: string;
     jsonschema: any;
+    options?: ReactQueryOptions
```
Comment (pyramation) [inline packages/wasm-ast-types/src/react-query.ts:212]: this is cool! I like the options object :)

File: packages/wasm-ast-types/src/react-query.spec.ts [hunk 5]
```diff
@@ -57,5 +57,12 @@ it('createReactQueryHooks', () => {
       'Sg721',
       'Sg721QueryClient'
     )))
+  expectCode(t.program(
+    createReactQueryHooks(
+      query_msg,
+      'Sg721',
+      'Sg721QueryClient',
+      { optionalClient: true }
+    )))
```
Comment (pyramation) [inline packages/wasm-ast-types/src/react-query.spec.ts:66]: I think `createReactQueryHooks` takes an object?

Maybe we can replace this test case with:

```js
describe('createReactQueryHooks', () => {
  it('no options', () => {
    expectCode(t.program(
      createReactQueryHooks({
        queryMsg: query_msg,
        contractName: 'Sg721',
        QueryClient: 'Sg721QueryClient'
      })));
  });
  it('optionalClient: false', () => {
    expectCode(t.program(
      createReactQueryHooks({
        queryMsg: query_msg,
        contractName: 'Sg721',
        QueryClient: 'Sg721QueryClient',
        options: { optionalClient: false }
      })));
  })
});

```

**Review by pyramation — CHANGES_REQUESTED** [review]
this is great! thanks for doing everything so quickly. I made a few other notes. One final thought, let's make at the top of the `createReactQueryHooks` call, a way to create default options. 

if there are no options passed in, or `options` is null, let's set it to a default object = `{ optionalClient: false }`. This way current projects that use the library can expect the same result. 

I just tested and this should do the trick:

```js
export const createReactQueryHooks = ({
    queryMsg,
    contractName,
    QueryClient,
    options = { optionalClient: false }
}: ReactQueryHooks) => {
...
```

File: packages/wasm-ast-types/src/react-query.ts (hunk 3 above)
Comment (adairrr) [reply packages/wasm-ast-types/src/react-query.ts:170]: Yep! that's what I meant by 
> Can add a ternary here as well if desired! 
:P

File: packages/wasm-ast-types/src/react-query.spec.ts (hunk 5 above)
Comment (adairrr) [reply packages/wasm-ast-types/src/react-query.spec.ts:66]: It's optional so it works without the parameter but I can make it required.

Comment (pyramation) [reply packages/wasm-ast-types/src/react-query.spec.ts:66]: oh — we can make `options` optional — but just give it defaults like I was mentioning before.

Comment (pyramation) [reply packages/wasm-ast-types/src/react-query.spec.ts:66]: like this

```js
export const createReactQueryHooks = ({
    queryMsg,
    contractName,
    QueryClient,
    options = { optionalClient: false }
}: ReactQueryHooks) => {
```

**Comment by pyramation** [comment]
- [ ] let's add to each one of these `packages/cosmwasm-typescript-gen/__tests__/cosmwasmts.test.ts`

```diff
it('sg721', async () => {
    const out = OUTPUT_DIR + '/sg721';
    const schemaDir = FIXTURE_DIR + '/sg721/';

    const schemas = readSchemas({ schemaDir, argv: {} });
    await generate('Sg721', schemas, out);
    await fromPartial('Sg721', schemas, out);
    await recoil('Sg721', schemas, out);
    await reactQuery('Sg721', schemas, out);
+  await reactQuery('Sg721', schemas, out, { optionalClient: true });
})
```

File: packages/wasm-ast-types/src/react-query.spec.ts (hunk 5 above)
Comment (adairrr) [reply packages/wasm-ast-types/src/react-query.spec.ts:66]: Sure. Now it's an object. Funny enough I did and then undid this [here](https://github.com/CosmWasm/ts-codegen/pull/31/commits/5c5d9b1a3e0a7eca2c9af43938090723442b9c35#diff-4b3303ef14bc8d8253c59270a7611661ef35dbbb1b3df79b1360dd7160610901)

**Comment by pyramation** [comment]
hey @NoahSaso can you check this out for your use case? is there something else we need to add?

**Comment by pyramation** [comment]
thanks for the awesome work!

```
Successfully published:
 - @cosmwasm/ts-codegen@0.5.0
 ```

**Pull request merged**